### PR TITLE
Dialogue translation pipeline

### PR DESCRIPTION
**Pull request by Mehrad0711**
Accompanying https://github.com/stanford-oval/thingpedia-common-devices/pull/444

### CODE REVIEW

File: pipelines/translation.py
Comment (gcampax) [inline pipelines/translation.py:49]: Production models need bootleg, fewshot, calibration, ood. We need a pipeline with those on.
Btw, does bootleg work with non-English data? Do we need to train a Bootleg against non-English wikipedia?

In reality, I wonder if we should have a mega-pipeline that has all the languages we care about, with Generate (shared with English), and then one branch of each model for each language. It might be expensive though.

File: pipelines/translation.py
Comment (gcampax) [inline pipelines/translation.py]: Let's call this translation model? Also, why do I have to specify both "marian" and the actual model name? Can we guess one from the other?

File: pipelines/training.py
Comment (gcampax) [inline pipelines/training.py:856]: How do you handle fewshot for translation?

File: pipelines/training.py
Comment (gcampax) [inline pipelines/training.py:794]: I think OOD should come after translation, because we need to distinguish between translated commands and translated OODs.

File: pipelines/training.py
Comment (gcampax) [inline pipelines/training.py:858]: I think paraphrasing should also come before translation, because we don't have a paraphraser in the target language. But you experimented with this before, which way does it work best?

(I'm not using paraphrasing right now in production because it's not worth the compute cost, but might reconsider in the future)

File: pipelines/translation.py
Comment (Mehrad0711) [reply pipelines/translation.py:49]: Just to be clear this is not a production-ready pipeline! This is minimum infra to get an NLU model to test. We should discuss what needs to be done to get there, but I wouldn't be able to work on all very soon.
Also deploying it with all the extra stuff now is expensive as you noted and we need to be confident the rest of the system works in another language. Then we can start adding these additional pieces to improve accuracy.

Multilingual bootleg is a project by itself. The challenge is collecting clean data: a lot of text with mentions and their qids. Multilingual dumps are available but anchor texts are often misaligned and noisy. I saw some papers doing cross-lingual entity linking without the need to retrain from scratch. Might worth having someone play around with those.

File: pipelines/translation.py
Comment (Mehrad0711) [reply pipelines/translation.py]: Sure I can change "model_name_or_path" to "translation_model".
"nmt" is equivalent to what we have as "model" in the train pipeline. The user chooses it. It's servers as a quick clue of the translation details: model and generation params. For example "mbart_m2o_temp0.2_penalty0.3"

File: pipelines/training.py
Comment (Mehrad0711) [reply pipelines/training.py:856]: First, we need to collect some fewshot data for foreign languages.
Then we can have multiple annotated files in thingpedia and change makefile to parse the correct one according to locale.

File: pipelines/training.py
Comment (Mehrad0711) [reply pipelines/training.py:794]: We can do OOD for other languages too although it's tricky. Your English ood is a mix of logs and all the data from staging. We can translate logs once and cache it. Staging data needs a round of translation for every experiment which can be expensive if we do for each run. We can do it once and sync and cache that and update it every now and then when staging files get updated.

File: pipelines/training.py
Comment (Mehrad0711) [reply pipelines/training.py:858]: I never got to do a comparison between training on translated English paraphrases vs. paraphrases in target language. 
We can get 1-2% improvement if we try very hard. My feeling is paraphrasing won't matter as much for translated data cause translation by itself is a form of paraphrasing (it maintains semantics but changes language) Although I have a pipeline for round-trip and masking paraphrasing already so we can have someone test different strategies and see what works best for the product model.

File: pipelines/translation.py
Comment (gcampax) [reply pipelines/translation.py]: Oh so it's a user chosen ID. Why is this not keyed off the normal `model` field?

File: pipelines/training.py
Comment (gcampax) [reply pipelines/training.py:856]: Right. Ugh.

File: pipelines/training.py
Comment (gcampax) [reply pipelines/training.py:858]: Yeah we don't have the resources to do that right now. Let's not worry about paraphrasing.

File: pipelines/translation.py
Comment (gcampax) [reply pipelines/translation.py:49]: It will be difficult to use for anything if it's not going to give us good accuracy, at least in one skill. We definitely need some way to validate on native language data, and some fewshot. This is going to be tricky.

**Review by gcampax — APPROVED** [review]
Lets merge this with those minor changes to the arg names for now, and we'll figure out what to do about prod later.

File: pipelines/translation.py
Comment (gcampax) [reply pipelines/translation.py]: Sure.

File: pipelines/translation.py
Comment (Mehrad0711) [reply pipelines/translation.py:49]: It's testable but the accuracy would be 70% instead of 75% for example.
It's the same chicken-and-egg problem as English. We need to start deploying sth to get user logs.
We can also crowd-source the data if we don't mind paying.

**Comment by Mehrad0711** [comment]
Lemme test the pipeline again before merging to make sure the arg changes, etc. didn't break things.

File: pipelines/translation.py
Comment (gcampax) [reply pipelines/translation.py:49]: I don't trust testing on translated data. But anyway let's not worry about it for now. We need to bootstrap first.

File: pipelines/translation.py
Comment (Mehrad0711) [reply pipelines/translation.py]: The "model" indicates training model. "nmt" indicates NMT model. They are different. You may use different "model"s to train different models on the same translated dataset so "nmt" wouldn't change.
Also you may just run translation without training.

Comment (gcampax) [reply pipelines/translation.py]: That's what the "dataset" parameter is then. Like, if it is used just to make S3 directories to write into, let's make it "${dataset}-nmt-${tgt_lang}" and be done with it. One fewer parameter to type.

Comment (Mehrad0711) [reply pipelines/translation.py]: Ok final call.
I would replace "model_name_or_path" with "translation_model" and
"nmt" with "nmt_id". Sounds good?

Comment (gcampax) [reply pipelines/translation.py]: Agree on "translation_model". I just don't understand why you keep the extra parameter for the dataset ID there though?

File: pipelines/translation.py
Comment (Mehrad0711) [reply pipelines/translation.py:49]: yeah so it's either user logs or crowd-sourcing to get real data.

File: pipelines/translation.py
Comment (Mehrad0711) [reply pipelines/translation.py]: It's servers a similar purpose as "model" for training except this is for translation. Honestly, don't worry about it. I can give it a default value if you don't wanna be bothered by what to pass here every time.

**Comment by Mehrad0711** [comment]
Ok finally I got a clean run on k8s but two changes (1)https://github.com/stanford-oval/thingpedia-common-devices/pull/448 (2)https://github.com/stanford-oval/genie-toolkit/pull/824 are needed before we can merge this.

**Pull request merged**
